File: Elixir/450.exs
# Doesnt work. I will correct it futurely

defmodule TreeNode do
  @type t :: %__MODULE__{
    val: integer,
    left: TreeNode.t() | nil,
    right: TreeNode.t() | nil
  }
  defstruct val: 0, left: nil, right: nil
end

defmodule Solution do
  @spec delete_node(root :: TreeNode.t | nil, key :: integer) :: TreeNode.t | nil
  def delete_node(root, key) do
    get_node(root, key)
    |> del_node(key)
    root
  end

  defp get_node(node, key) do
    cond do
      nil == node -> nil
      key == node.val -> node
      key <  node.val -> get_node(node.left,  key)
      key >  node.val -> get_node(node.right, key)
    end
  end

  defp del_node(nil, _), do: nil
  defp del_node(node, key) do
    case {node.left, node.right} do
      {nil,   nil}  -> node = nil
      {left,  nil}  -> node = node.left
      {nil, right}  -> node = node.right
      {left, right} -> node = swap_for_successor(node)
    end
  end

  defp swap_for_successor(node) do
    succ = successor(node, node.right)
    node = %{node | right: succ.right}
  end

  defp successor(node, nil),  do: node
  defp successor(node, left), do: successor(node.left, left.left)
end

File: Python/450.py
# Delete Node in a BST
# Medium

# This still doesnt work

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def deleteNode(root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        
    def get_node(node, key, parent=None):
        if node == None: return None
        elif key < node.val: return get_node(node.left,  key, node)
        elif key > node.val: return get_node(node.right, key, node)
        else: return (node, parent)
        
    def del_node(node, parent) -> None:
        if parent is None:
            node = None
            return 
        if parent.left == node:
            if node.right is not None:
                parent.left = node.right
            else: parent.left = node.left
        else: 
            if node.right is not None:
                parent.right = node.right
            else: parent.right = node.left

    def del_twin_child_node(node) -> None:
        sparent = node
        succ = node.right
        while True:
            if succ.left is not None:
                sparent = succ
                succ = succ.left
            else: break
        del_node(succ, sparent)
        node.val = succ.val

    t = get_node(root, key)
    if t is None: return root
    
    remove_me, parent = t
    if   remove_me.left is None and remove_me.right is None:
        del_node(remove_me, parent)
    elif remove_me.left is not None and remove_me.right is not None:
        del_twin_child_node(remove_me)
    else:
        del_node(remove_me, parent)

    return root